src/main/java/jenkins/bouncycastle/api/PEMManager.java

```diff
@@ -0,0 +1,403 @@
+     * from the current PEM data.
+     * @throws IOException thrown if a problem exists creating the fingerprint
+     */
+    @CheckForNull
+    public String getPrivateKeyFingerprint() throws IOException {
+        PrivateKey key = toPrivateKey();
+        if (key == null) {
+            return null;
+        }
+        return hexEncode(getKeyDigestSHA1(key));
+    }
+
+    /**
+     * Obtains the fingerprint of the public key in the "ab:cd:ef:...:12" format, which basically is an MD5 digest from
+     * the key, encoded in hex format.
+     * 
+     * @return public key fingerprint in hex format "ab:cd:ef:...:12", null if the public key could not be obtained from
+     * the current PEM data.
+     * @throws IOException if a problem exists creating the fingerprint
+     */
+    @CheckForNull
+    public String getPublicKeyFingerprint() throws IOException {
+        PublicKey key = toPublicKey();
+        if (key == null) {
+            return null;
+        }
+        return hexEncode(getKeyDigestMD5(key));
+    }
+
+    /**
+     * Generates an SHA1 digest from a Key object
+     * 
+     * @param k the key to generate the digest from
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
+     */
+    @Nonnull
+    public static byte[] getKeyDigestSHA1(@Nonnull Key k) throws IOException {
+        try {
+            return getKeyDigest(k, "SHA1");
+        } catch (NoSuchAlgorithmException e) {
+           throw new AssertionError("SHA1 algorithm not found to create digest");
+        }
+    }
+
+    /**
+     * Generates an MD5 digest from a Key object
+     * 
+     * @param k the key to generate the digest from
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
+     */
+    @Nonnull
+    public static byte[] getKeyDigestMD5(@Nonnull Key k) throws IOException {
+        try {
+            return getKeyDigest(k, "MD5");
+        } catch (NoSuchAlgorithmException e) {
+            throw new AssertionError("SHA1 algorithm not found to create digest");
+        }
+    }
+
+    /**
+     * Generates an digest from a Key object in the specified digest format. The supported digest formats will depend on
+     * the JVM API.
+     * 
+     * @param k key to generate the digest from
+     * @param algorithm digest format
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
+     * @throws NoSuchAlgorithmException when provided digest algorithm is not available
+     */
+    @Nonnull
+    public static byte[] getKeyDigest(@Nonnull Key k, @Nonnull String algorithm) throws IOException, NoSuchAlgorithmException {
+        MessageDigest md5 = MessageDigest.getInstance(algorithm);
+
+        DigestInputStream in = new DigestInputStream(new ByteArrayInputStream(k.getEncoded()), md5);
```

Comment:
inefficient.

```
MessageDigest md = MessageDigest.getInstance(algorithm);
md.update(k.getEncoded());
return md.digest();
```

then you have no intermediate garbage and you can loose the `IOException`